x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -115,96 +165,127 @@ private void loadModel(String modelId) {
+        }
+    }
+
+    private void cacheEvictionListener(RemovalNotification<String, LocalModel> notification) {
+        if (notification.getRemovalReason() == RemovalNotification.RemovalReason.EVICTED) {
+            String msg = new ParameterizedMessage(
+                "model cache entry evicted." +
+                    "current cache [{}] current max [{}] model size [{}]. " +
+                    "If this is undesired, consider updating setting [{}] or [{}].",
+                new ByteSizeValue(localModelCache.weight()).getStringRep(),
+                maxCacheSize.getStringRep(),
+                new ByteSizeValue(notification.getValue().ramBytesUsed()).getStringRep(),
+                INFERENCE_MODEL_CACHE_SIZE.getKey(),
+                INFERENCE_MODEL_CACHE_TTL.getKey()).getFormattedMessage();
+            auditIfNecessary(notification.getKey(), msg);
         }
     }
 
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
-        if (event.changedCustomMetaDataSet().contains(IngestMetadata.TYPE)) {
-            ClusterState state = event.state();
-            IngestMetadata currentIngestMetadata = state.metaData().custom(IngestMetadata.TYPE);
-            Set<String> allReferencedModelKeys = getReferencedModelKeys(currentIngestMetadata);
-            // The listeners still waiting for a model and we are canceling the load?
-            List<Tuple<String, List<ActionListener<Model>>>> drainWithFailure = new ArrayList<>();
-            synchronized (loadingListeners) {
-                HashSet<String> loadedModelBeforeClusterState = logger.isTraceEnabled() ? new HashSet<>(loadedModels.keySet()) : null;
-                HashSet<String> loadingModelBeforeClusterState = logger.isTraceEnabled() ? new HashSet<>(loadingListeners.keySet()) : null;
-                // If we had models still loading here but are no longer referenced
-                // we should remove them from loadingListeners and alert the listeners
-                for (String modelId : loadingListeners.keySet()) {
-                    if (allReferencedModelKeys.contains(modelId) == false) {
-                        drainWithFailure.add(Tuple.tuple(modelId, new ArrayList<>(loadingListeners.remove(modelId))));
-                    }
-                }
+        // If ingest data has not changed or if the current node is not an ingest node, don't bother caching models
+        if (event.changedCustomMetaDataSet().contains(IngestMetadata.TYPE) == false ||
+            event.state().nodes().getLocalNode().isIngestNode() == false) {
+            return;
+        }
 
-                // Remove all cached models that are not referenced by any processors
-                loadedModels.keySet().retainAll(allReferencedModelKeys);
+        ClusterState state = event.state();
+        IngestMetadata currentIngestMetadata = state.metaData().custom(IngestMetadata.TYPE);
+        Set<String> allReferencedModelKeys = getReferencedModelKeys(currentIngestMetadata);
```

Comment:
what if the cluster state update does not change anything w.r.t. models, which is probably mostly the case? I miss a shortcut optimization.

Reply:
nice catch, should exit early if `allReferenceModelKeys. equalTo(referencedModels)`